### DISCUSSION TITLE
How to properly reference data added using the CodeAssembler .db function in 32-bit mode?

### DISCUSSION THREAD

**Discussion by netadr** [icedland/iced#279] [Q&A] [answered]
I've been having an issue referencing data embedded using the `.db`  function of `CodeAssembler`:

Here's a snippet that demonstrates the issue:
```rust
let mut a = CodeAssembler::new(32)?;
let mut skip_data = a.create_label();
let mut data = a.create_label();

a.jmp(skip_data)?;
a.set_label(&mut data)?;
a.db(b"\x90\xCC\xF1\x90")?;
a.set_label(&mut skip_data)?;
a.lea(eax, dword_ptr(data))?;

a.assemble(0x0)?;
```
Calling `.assemble` returns the following error:
`Error: Operand 1: 64-bit addressing can only be used in 64-bit mode : 0x1`

Is this a case of user error or a bug within the library?

**Comment by wtfsck** [comment, marked as answer]
That example is from the README and uses 64-bit mode and will fail in 16/32-bit mode.

One workaround is to get the address of the label after assembling it (see `assemble_options()?.label_ip()` and then patching an instruction, eg. a 'mov eax,imm32' instruction. You can get the offset of the imm32 by passing in `BlockEncoderOptions::RETURN_CONSTANT_OFFSETS | BlockEncoderOptions::RETURN_NEW_INSTRUCTION_OFFSETS` to `assemble_options()`.

Another one is to do something like:

```asm
call there
my_data:
db 1,2,3,4
there:
pop eax
```

`eax` now has the address of `my_data`.

I created an issue: https://github.com/icedland/iced/issues/281
